File: frontend/html/scripts/protocol/client/messages/ClientMessageUpdateCanvas.js
import ClientMessageType from "../ClientMessageType.js";
import ClientMessage from "../ClientMessage.js";

export default class ClientMessageUpdateCanvas extends ClientMessage {
    constructor(canvasObjects) {
        super(ClientMessageType.UPDATE_CANVAS);
        this.canvasObjectWrappers = canvasObjects;//CanvasObjectWrapper
    }

    serialize(writer) {
        super.serialize(writer);
        writer.writeUint8(this.canvasObjectWrappers.size);
        this.canvasObjectWrappers.forEach((map, canvasObjectType) => {
            writer.writeUint8(canvasObjectType);
            writer.writeUint16(map.size);
            map.forEach((canvasObjectWrapper, id) => {
                writer.writeInt16(id);
                canvasObjectWrapper.serialize(writer);
            });
        });
    }
}

File: board-room/src/main/java/net/stzups/board/room/server/websocket/protocol/client/messages/ClientMessageUpdateCanvas.java
package net.stzups.board.room.server.websocket.protocol.client.messages;

import io.netty.buffer.ByteBuf;
import net.stzups.board.data.objects.canvas.object.CanvasObjectType;
import net.stzups.board.data.objects.canvas.object.CanvasObjectWrapper;
import net.stzups.board.room.server.websocket.protocol.client.ClientMessage;
import net.stzups.board.room.server.websocket.protocol.client.ClientMessageType;

import java.util.HashMap;
import java.util.Map;

public class ClientMessageUpdateCanvas extends ClientMessage {
    private final Map<CanvasObjectType, Map<Short, CanvasObjectWrapper>> canvas = new HashMap<>();

    public ClientMessageUpdateCanvas(ByteBuf byteBuf) {
        super(ClientMessageType.UPDATE_CANVAS);
        int length = byteBuf.readUnsignedByte();
        for (int i = 0; i < length; i++) {
            CanvasObjectType canvasObjectType = CanvasObjectType.valueOf(byteBuf.readUnsignedByte());
            Map<Short, CanvasObjectWrapper> map = new HashMap<>();
            int lengthJ = byteBuf.readUnsignedShort();
            for (int j = 0; j < lengthJ; j++) {
                map.put(byteBuf.readShort(), new CanvasObjectWrapper(canvasObjectType, byteBuf));
            }
            canvas.put(canvasObjectType, map);
        }
    }

    public Map<CanvasObjectType, Map<Short, CanvasObjectWrapper>> getCanvasObjects() {
        return canvas;
    }
}
